### PR TITLE
STOR-1334: update storage operator to read featuregates from API on standalone OCP

### PR DESCRIPTION
**Pull request by deads2k**
The previous method of using the featuregate object and a potentially out of date hardcoded list is left for hypershift until hypershift is ready to switch.

This allows coordination of featuregates via openshift/api and vendoring to openshift/cluster-config-operator. Every operator on standalone kcp consumes from `oc get featuregates` status.

### CODE REVIEW

**Comment by jsafrane** [comment]
cc @openshift/storage

File: pkg/operator/standalone_starter.go [hunk 1]
```diff
@@ -33,6 +35,31 @@ func startControllerStandAlone(ctx context.Context, controllerConfig *controller
...
 
+	desiredVersion := status.VersionForOperatorFromEnv()
+	missingVersion := "0.0.1-snapshot"
+
+	// By default, this will exit(0) the process if the featuregates ever change to a different set of values.
+	featureGateAccessor := featuregates.NewFeatureGateAccess(
+		desiredVersion, missingVersion,
+		clients.ConfigInformers.Config().V1().ClusterVersions(), clients.ConfigInformers.Config().V1().FeatureGates(),
+		controllerConfig.EventRecorder,
+	)
```
Comment (jsafrane) [inline pkg/operator/standalone_starter.go:46]: Will `exit(0)` trigger a new leader election? That will slow down feature enablement considerably, without any indication that the operator is Progressing.

**Comment by jsafrane** [comment]
/lgtm
/approve

If we find out that `exit()` is bad, we can always refactor it back to Informer.

**Comment by jsafrane** [comment]
Note to QE: we should test that this PR does not introduce any regression in how SharedResource CSI driver operator is started after `TechPreviewNoUpgrade` is set. Apart from longer delay between setting TechPreviewNoUpgrade and the driver installed, few minutes max.

**Comment by jsafrane** [comment]
/retitle STOR-1334: update storage operator to read featuregates from API on standalone OCP

Adding a story for visibility on our board for QE.

**Comment by duanwei33** [comment]
QE are starting pre-merge test.

**Comment by duanwei33** [comment]
Simply verified with the PR:
1. When I enable the TechPreviewNoUpgrade, we see enabled featuregates are listed as below:
```
$ oc get featuregate cluster -o yaml
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  creationTimestamp: "2023-05-16T09:56:43Z"
  generation: 2
  name: cluster
  resourceVersion: "65285"
  uid: 164d3780-364d-49f1-bf9e-e33c67c94397
spec:
  featureSet: TechPreviewNoUpgrade
status:
  featureGates:
  - enabled:
    - name: AdmissionWebhookMatchConditions
    - name: AzureWorkloadIdentity
    - name: BuildCSIVolumes
    - name: CSIDriverSharedResource
    - name: DynamicResourceAllocation
    - name: ExternalCloudProvider
    - name: ExternalCloudProviderAzure
    - name: ExternalCloudProviderGCP
    - name: InsightsConfigAPI
    - name: MachineAPIProviderOpenStack
    - name: MatchLabelKeysInPodTopologySpread
    - name: NodeSwap
    - name: OpenShiftPodSecurityAdmission
    - name: PDBUnhealthyPodEvictionPolicy
    - name: RetroactiveDefaultStorageClass
    version: 4.14.0-0.test-2023-05-16-094624-ci-ln-5c4g47k-latest
```
2. SharedResource CSI Driver is installed successfully:
```
$ oc get clustercsidrivers csi.sharedresource.openshift.io
NAME                              AGE
csi.sharedresource.openshift.io   25m

$ oc -n openshift-cluster-csi-drivers get pod | grep share
shared-resource-csi-driver-node-mxgs6                  2/2     Running   0          21m
shared-resource-csi-driver-node-r2b2h                  2/2     Running   2          21m
shared-resource-csi-driver-node-rblr4                  2/2     Running   2          21m
shared-resource-csi-driver-operator-5b64bd68c4-d4wr7   1/1     Running   0          15m
shared-resource-csi-driver-webhook-d586db8bb-gjbkh     1/1     Running   0          19m
shared-resource-csi-driver-webhook-d586db8bb-hqwqj     1/1     Running   0          15m
```

File: pkg/operator/csidriveroperator/driverstarter_standalone.go [hunk 2]
```diff
@@ -269,7 +268,8 @@ func shouldRunController(cfg csioperatorclient.CSIOperatorConfig, infrastructure
 		return true, nil
 	}
 
-	if !featureGateEnabled(fg, cfg.RequireFeatureGate) {
+	knownFeatures := sets.New[configv1.FeatureGateName](fg.KnownFeatures()...)
```
Comment (sjenning) [inline pkg/operator/csidriveroperator/driverstarter_standalone.go:271]: This is panicing on the hypershift control plane and blocking ci release stream

```
E0530 00:05:42.329947       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 820 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x26a62a0?, 0x475c5c0})
	k8s.io/apimachinery@v0.27.1/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0xc000699408, 0x1, 0x2545ca0?})
	k8s.io/apimachinery@v0.27.1/pkg/util/runtime/runtime.go:49 +0x75
panic({0x26a62a0, 0x475c5c0})
	runtime/panic.go:884 +0x213
github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator.shouldRunController({{0x2b15110, 0x1f}, {0x2ade262, 0x6}, {0x2ae8cf2, 0xc}, 0x0, {0x0, 0x0, 0x0}, ...}, ...)
	github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator/driverstarter_standalone.go:271 +0x157
```

**Pull request merged**